### PR TITLE
Optimize `T? ?? <expr>` to `T.GetValueOrDefault(<expr>)` for a side-effectless `<expr>` where `T?` is a nullable value type

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Closes https://github.com/dotnet/roslyn/issues/56007

### CODE REVIEW

**Review by Youssef1313 — COMMENTED** [review]
Might be quite overlapping with https://github.com/dotnet/roslyn/pull/68069

cc @jjonescz

**Comment by DoctorKrolic** [comment]
Most likely yes, as this PR changes code in the same file as #68069, so 1 of us is probably gonna have merge conflicts if the other PR is merged. However, optimizations themselves are different, so both PRs are still valuable

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs
Comment (Rekkonnect) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:1865]: 'present'

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs
Comment (Rekkonnect) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs:149]: Perhaps extract
```csharp
rewrittenRight.Type.Equals(leftUnderlyingType, TypeCompareKind.AllIgnoreOptions)
```
in an outer if, but also preserve evaluation of `IsDefaultValue` and `ConstantValueOpt is not null` first?

**Review by Rekkonnect — APPROVED** [review]
Nice one with a few nits

**Comment by DoctorKrolic** [comment]
Can I have a review here please? PR is 2 weeks old, but still no attention from the compiler team :(

**Comment by Rekkonnect** [comment]
It seems that it's milestoned for 17.9, which is far away from the current version in preview, 17.8.

To make you feel better I've had open PRs for months, also milestoned for that version, so I'm guessing attention will be given sooner or later.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:1872]: The found member might still be bad

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_BinaryOperator.cs:1868]: >TryGetNullableMethodSoft [](http://example.com/codeflow?start=21&length=24)

I recommend to not "reinvent a wheel" and use one of the existing helpers. The Binder probably already has a helper with desired behavior

**Comment by AlekseyTs** [comment]
Consider implementing the same improvement for VB

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs:138]: >var rightIsOfUnderlyingTypeOfLeft = rewrittenRight.Type.Equals(leftUnderlyingType, TypeCompareKind.AllIgnoreOptions); [](http://example.com/codeflow?start=16&length=117)

It looks like this local and `leftUnderlyingType` are not necessary. We can simply restore the type comparison condition where it used to be and in its original form.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_NullCoalescingOperator.cs:143]: >TryGetNullableMethodSoft [](http://example.com/codeflow?start=20&length=24)

Are we testing scenario when this method isn't found?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:5965]: >// (5,16): error CS0656: Missing compiler required member 'System.Nullable`1.GetValueOrDefault' [](http://example.com/codeflow?start=16&length=95)

The error looks surprising. Why can't we simply emit code like in the test below?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:5965]: I see, it also relies on GetValueOrDefault.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 11)

**Comment by DoctorKrolic** [comment]
Implemented same change for VB side. VB code is structured a bit differently, so I didn't implement such a gracefull fallback like in C# but it probably doesn't matter much since it is not expected to ever encounter such errors in usual day-to-day life.

I also extended the optimization to work for locals and parameters. This changes the order of evaluation, but I cannot think of a case, where it can lead to unexpected behavior of a compiled program. Ci will double-check that)

**Comment by AlekseyTs** [comment]
> I also extended the optimization to work for locals and parameters

Consider adjusting PR title accordingly.

**Comment by AlekseyTs** [comment]
>             if (useSiteInfo.DiagnosticInfo != null)

It doesn't look like behavior around use site errors is consistent with ```GetWellKnownTypeMember```, especially with respect to the ```isOptional``` parameter.

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:1680 in 9ad8c58. [](commit_id = 9ad8c5847f5e83acef9213c05f816fca5dbfa317, deletion_comment = False)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:5019]: An attempt to dereference might cause an exception. We probably shouldn't optimize scenarios like this.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:4980]: >x ?? y [](http://example.com/codeflow?start=60&length=6)

It looks like we are missing a test for a scenario with `y` as a local.

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_ConditionalExpressions.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_ConditionalExpressions.vb:244]: >Return NullableValueOrDefault(rewrittenLeft, rewrittenRight) [](http://example.com/codeflow?start=20&length=60)

We shouldn't report an error when this helper is missing or otherwise bad

File: src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTests.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenTests.vb:3663]: Is this a breaking change?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 17)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:5019]: Ok, will exclude all non-ref locals/parameters from this optimization. But out of curiosity, in what cases dereferencing throws?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:5019]: >  in what cases dereferencing throws?

An access violation, for example.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:5019]: Did you mean "exclude all ref locals/parameters"?

Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenOperators.cs:5019]: > Did you mean "exclude all ref locals/parameters"?

I had 2 thoughts at that moment: "include only non-ref locals/parameters" and "exclude all ref locals/parameters". These 2 thoughts collided and we got a merge conflict)

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_ConditionalExpressions.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_ConditionalExpressions.vb:244]: >(rewrittenRight.Kind = BoundKind.Local AndAlso Not DirectCast(rewrittenRight, BoundLocal).LocalSymbol.IsRef) [](http://example.com/codeflow?start=19&length=108)

This isn't a request for a change. Just a note. This expression could be shortened to:
```
Not If(TryCast(rewrittenRight, BoundLocal)?.LocalSymbol.IsRef, True)
```

**Comment by AlekseyTs** [comment]
Done with review pass (commit 19). The following threads are still not resolved:
- https://github.com/dotnet/roslyn/pull/70199#issuecomment-1764606158
- https://github.com/dotnet/roslyn/pull/70199#pullrequestreview-1680172938
- https://github.com/dotnet/roslyn/pull/70199#pullrequestreview-1680591137

**Comment by AlekseyTs** [comment]
@DoctorKrolic It looks like there are legitimate test failures.

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:1695]: >diagnosticInfo: null [](http://example.com/codeflow?start=66&length=20)

Is it safe to call `ReportUseSiteDiagnostic` with null as an input below? Even if it is safe, it looks like creating a new `UseSiteInfo` instance and then calling `ReportUseSiteDiagnostic` is not useful. Instead, consider restructuring the code to call `ReportUseSiteDiagnostic` only for non-optional scenario.

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_LateBindingHelpers.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_LateBindingHelpers.vb:994]: >isOptional [](http://example.com/codeflow?start=116&length=10)

I suggest simply following the pattern used in the function above and reverting all changes made in `Binder.vb`.

File: src/Compilers/VisualBasic/Portable/Microsoft.CodeAnalysis.VisualBasic.vbproj
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Microsoft.CodeAnalysis.VisualBasic.vbproj:77]: > Is it that much necessary to revert this change?

Yes. This is a separate issue and it doesn't look like it blocks the merge for the optimization. If you think there is a problem with the project file, consider opening a separate issue with a clear explanation. Perhaps with VS screen shots because, for example, "polluting solution explorer with a "ghost" item" means nothing specific to me. Then, if you would like to propose a fix, submit a separate PR with it.

File: src/Compilers/VisualBasic/Portable/Binding/Binder.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/Binder.vb:570]: >If severity = DiagnosticSeverity.Error Then [](http://example.com/codeflow?start=24&length=43)

I think this check isn't necessary in VB because  `useSiteInfo.DiagnosticInfo` can be only an error in VB.

File: src/Compilers/VisualBasic/Portable/Binding/Binder.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Binding/Binder.vb:576]: >useSiteInfo = New UseSiteInfo(Of AssemblySymbol)(diagnosticInfo:=Nothing, useSiteInfo.PrimaryDependency, useSiteInfo.SecondaryDependencies) [](http://example.com/codeflow?start=24&length=139)

I do not think this code is reachable

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_LateBindingHelpers.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_LateBindingHelpers.vb:997]: >Return False [](http://example.com/codeflow?start=16&length=12)

This looks like a wrong thing to do for a non-optional case

**Comment by AlekseyTs** [comment]
Done with review pass (commit 24)

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_NullableHelpers.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_NullableHelpers.vb:204]: >NullableValueOrDefaultOpt [](http://example.com/codeflow?start=25&length=25)

I suggest adjusting behavior of the function above based on a new parameter. Otherwise, we are missing handling for scenarios that simply need unwrapping constructor or conversion

**Comment by AlekseyTs** [comment]
Done with review pass (commit 29)

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_NullableHelpers.vb
Comment (DoctorKrolic) [reply src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_NullableHelpers.vb]: There was already an overload that took 1 boolean parameter, so I had to rename it first

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_NullableHelpers.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_NullableHelpers.vb:181]: >NullableValueOrDefault(whenNotNull, isOptional) [](http://example.com/codeflow?start=59&length=47)

Simply inlining the value doesn't look appropriate, We don't want to create conditional access with null value, we want to return null value instead.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 31)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 32)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review.

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:1688]: Why don't we want to report warnings in this case? It seems likely to me that this will cause missed use site diagnostics/tracking somewhere.

File: src/Compilers/CSharp/Test/Symbol/Symbols/MissingSpecialMember.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Symbol/Symbols/MissingSpecialMember.cs:1593]: Seems like we should add a version of this test that marks both members missing.

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_ConditionalExpressions.vb
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_ConditionalExpressions.vb:245]:
```suggestion
                ' Optimize If(left, right) to left.GetValueOrDefault(right) when left is T? and right is a side-effectless expression of type T
```

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_LateBindingHelpers.vb
Comment (333fred) [inline src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_LateBindingHelpers.vb:997]: This is closer to the behavior I'd expect to see in C#; either we report the warning for the special member, or we just bail in the presence of any use-site diagnostic, regardless of whether it's an error.

**Review by 333fred — COMMENTED** [review]
Done review pass.

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:1688]: This is in sync with existing logic in `GetWellKnownTypeMember`, wich also swallows use site warnings in case the requested member is optional:
https://github.com/dotnet/roslyn/blob/30e683e2a6953a446a34ac5fc499287664061b96/src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs#L1805-L1818
TBH I don't fully understand what use site diagnostics even are, but it is intuitive that similar concepts should have similar behavior. And this change was suggested by Aleksey

File: src/Compilers/CSharp/Test/Symbol/Symbols/MissingSpecialMember.cs
Comment (DoctorKrolic) [reply src/Compilers/CSharp/Test/Symbol/Symbols/MissingSpecialMember.cs:1593]: I tried to keep all tests for this optimization close together in `CodeGenOperators.cs`. The test you are requesting here is already there and is named `TestNullCoalesce_NullableNonDefaultConstant_MissingGetValueOrDefaultAndGetValueOrDefaultWithADefaultValueParameter`
This test was fixed afterwards and I don't see much value in bringing other 'missing member' test cases to this file

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:1688]: > Why don't we want to report warnings in this case?

I think the reason is to avoid an unnecessary noise for methods that user didn't use explicitly. This is a preexisting behavior for well-known members.

>  It seems likely to me that this will cause missed use site diagnostics/tracking somewhere.

I cannot think of any reason why this would be the case. Do you have something specific in mind?

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_LateBindingHelpers.vb
Comment (DoctorKrolic) [reply src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_LateBindingHelpers.vb:997]: According to Aleksey, there are no use site warnings in VB. In such case "we just bail in the presence of any use-site diagnostic, regardless of whether it's an error" should be equvivalent to C#'s "bail on errors and ignore warnings" logic

Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_LateBindingHelpers.vb:997]: This is the pattern from the function right above this one, ```TryGetWellknownMember```.

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:1688]: > I cannot think of any reason why this would be the case. Do you have something specific in mind?

I am unsure whether we have any use-site diagnostics today that are just warnings. However, if we do, and if this method has such a warning, it's not going to be reported by the LocalRewriter machinery that creates the synthesized `BoundCall` here, and (though this is unlikely to matter in reality, especially given that this a SpecialType and not a WellKnownType) the use of the assembly symbol won't be tracked from this call.

> I think the reason is to avoid an unnecessary noise for methods that user didn't use explicitly

This argument doesn't make sense to me. Regardless of whether they intended to use it, they are using it, and it can have diagnostics. The pre-existing behavior for well-known members sounds like a bug.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:1688]: > if we do, and if this method has such a warning, it's not going to be reported by the LocalRewriter machinery that creates the synthesized BoundCall

That is the intended behavior that we already have for well-known members.

> the use of the assembly symbol won't be tracked from this call

Why is that? The use of assembly symbols is not tracked based on errors or warnings. And, if you carefully review the code of this function, you will notice that nothing about tracking used references is changed in this method. They are tracked exactly the same way they used to be tracked before.

> The pre-existing behavior for well-known members sounds like a bug.

I am happy to have this discussion outside of context of this PR. For now, I think we should simply follow what we were doing for well-known members. So, I suggest to not block this PR based on this.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:1688]: I'm ok to have the discussion outside the PR.

**Review by 333fred — APPROVED** [review]

**Comment by DoctorKrolic** [comment]
Can someone restart failed CI pls?

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 33)

**Comment by AlekseyTs** [comment]
@DoctorKrolic Thank you for the contribution!

**Pull request merged**
